File: web_api/api_mirai_http.py
import requests
import websocket
import json

port = 8080
host = 'localhost'
url = 'http://localhost:' + str(port)
ws_url = 'ws://localhost:' + str(port)
authKey = '123456789'
qq = 2155679839

class Send:

    def __init__(self):
        self.port = port
        self.url = url
        self.authKey = authKey
        self.sessionKey = ''
        self.qq = qq

        #认证，返回一个sessionKey
        self.auth()
        if self.sessionKey != '':
            self.verify()

    def about(self):
        r = requests.get(self.url + '/about')
        return json.loads(r.text)

    def auth(self):
        d = {'authKey': self.authKey}
        r = requests.post(self.url + '/auth', data = json.dumps(d))
        print(str(r.request))
        response = json.loads(r.text)
        if response['code'] == 0:
            self.sessionKey = response['session']
        else:
            print('web_api/api_mirai_http.py::Send::auth code:', response['code'])

    def verify(self):
        d = {
            'sessionKey': self.sessionKey,
            'qq': self.qq
            }
        r = requests.post(self.url + '/verify', data = json.dumps(d))
        response = json.loads(r.text)
        if response['code'] == 0:
            print('verify success')
        else:
            print('web_api/api_mirai_http.py::Send::verify code:', response['code'])

    def release(self):
        d = {
            'sessionKey': self.sessionKey,
            'qq': self.qq
            }
        r = requests.post(self.url + '/release', data = json.dumps(d))
        response = json.loads(r.text)
        if response['code'] == 0:
            pass
        else:
            print('web_api/api_mirai_http.py::Send::release code:', response['code'])

    #返回messageId
    def sendFriendMessage(self, qq, quote, messageChain):
        d = {
            'sessionKey': self.sessionKey,
            'qq': qq,
            'messageChain': messageChain
            }
        if quote != 0:
            d['quote'] = quote
        
        r = requests.post(self.url + '/sendFriendMessage', data = json.dumps(d))
        response = json.loads(r.text)

        if response['code'] == 0:
            return response['messageId']
        else:
            print('web_api/api_mirai_http.py::Send::sendFriendMessage code:', response['code'])
            return 0

    #返回messageId
    def sendTempMessage(self, group, qq, quote, messageChain):
        d = {
            'sessionKey': self.sessionKey,
            'qq': qq,
            'group': group,
            'messageChain': messageChain
            }
        if quote != 0:
            d['quote'] = quote
        
        r = requests.post(self.url + '/sendTempMessage', data = json.dumps(d))
        response = json.loads(r.text)

        if response['code'] == 0:
            return response['messageId']
        else:
            print('web_api/api_mirai_http.py::Send::sendTempMessage code:', response['code'])
            return 0

    #返回messageId
    def sendGroupMessage(self, group, quote, messageChain):
        d = {
            'sessionKey': self.sessionKey,
            'group': group,
            'messageChain': messageChain
            }
        if quote != 0:
            d['quote'] = quote
        #print(d, 'send')
        
        r = requests.post(self.url + '/sendGroupMessage', data = json.dumps(d))
        response = json.loads(r.text)

        if response['code'] == 0:
            return response['messageId']
        else:
            print('web_api/api_mirai_http.py::Send::sendGroupMessage code:', response['code'])
            return 0

    #TODO
    def sendImageMessage(self):
        pass

    #TODO
    def uploadImage(self):
        pass

    #撤回
    def recall(self, messageId):
        d = {
            'sessionKey': self.sessionKey,
            'target': messageId
            }
        
        r = requests.post(self.url + '/recall', data = json.dumps(d))
        response = json.loads(r.text)

        if response['code'] == 0:
            pass
        else:
            print('web_api/api_mirai_http.py::Send::recall code:', response['code'])
    
    #获取好友列表，返回json对象
    def getFriendList(self):
        d = {
            'sessionKey': self.sessionKey
            }
        
        r = requests.get(self.url + '/friendList', params = d)
        response = json.loads(r.text)

        return response

    #获取群列表，返回json对象
    def getGroupList(self):
        d = {
            'sessionKey': self.sessionKey
            }
        
        r = requests.get(self.url + '/groupList', params = d)
        response = json.loads(r.text)

        return response

    #获取群成员列表，返回json对象
    def getMemberList(self, group):
        d = {
            'sessionKey': self.sessionKey,
            'target': group
            }
        
        r = requests.get(self.url + '/memberList', params = d)
        response = json.loads(r.text)

        return response

    def muteAll(self, group):
        d = {
            'sessionKey': self.sessionKey,
            'target': group
            }
        
        r = requests.post(self.url + '/muteAll', data = json.dumps(d))
        response = json.loads(r.text)

        if response['code'] == 0:
            pass
        else:
            print('web_api/api_mirai_http.py::Send::muteAll code:', response['code'])

    def unmuteAll(self, group):
        d = {
            'sessionKey': self.sessionKey,
            'target': group
            }
        
        r = requests.post(self.url + '/unmuteAll', data = json.dumps(d))
        response = json.loads(r.text)

        if response['code'] == 0:
            pass
        else:
            print('web_api/api_mirai_http.py::Send::muteAll code:', response['code'])

    def mute(self, group, qq, time):
        d = {
            'sessionKey': self.sessionKey,
            'target': group,
            'memberId': qq,
            'time': time
            }
        
        r = requests.post(self.url + '/mute', data = json.dumps(d))
        response = json.loads(r.text)

        if response['code'] == 0:
            pass
        else:
            print('web_api/api_mirai_http.py::Send::mute code:', response['code'])

    def unmute(self, group, qq):
        d = {
            'sessionKey': self.sessionKey,
            'target': group,
            'memberId': qq
            }
        
        r = requests.post(self.url + '/unmute', data = json.dumps(d))
        response = json.loads(r.text)

        if response['code'] == 0:
            pass
        else:
            print('web_api/api_mirai_http.py::Send::unmute code:', response['code'])

    #TODO
    def kick(self):
        pass

    #bot退群
    def quit(self, group):
        d = {
            'sessionKey': self.sessionKey,
            'target': group
            }
        
        r = requests.post(self.url + '/quit', data = json.dumps(d))
        response = json.loads(r.text)

        if response['code'] == 0:
            pass
        else:
            print('web_api/api_mirai_http.py::Send::quit code:', response['code'])

    #TODO 修改群设置
    def setGroupConfig(self):
        pass

    #获取群设置
    def getGroupConfig(self, group):
        d = {
            'sessionKey': self.sessionKey,
            'target': group
            }
        
        r = requests.get(self.url + '/groupConfig', params = d)
        response = json.loads(r.text)

        return response

    #TODO 修改群员资料
    def setMemberInfo(self):
        pass

    #获取群员资料
    def getMemberInfo(self, group, qq):
        d = {
            'sessionKey': self.sessionKey,
            'target': group,
            'memberId': qq
            }
        
        r = requests.get(self.url + '/memberInfo', params = d)
        response = json.loads(r.text)

        return response

    #获取指定Session的配置
    def getSessionConfig(self):
        d = {
            'sessionKey': self.sessionKey
            }
        
        r = requests.get(self.url + '/config', params = d)
        response = json.loads(r.text)

        return response

    #设置指定Session的配置
    def setSessionConfig(self, cacheSize, enableWS):
        d = {
            'sessionKey': self.sessionKey,
            'cacheSize': cacheSize,
            'enableWebsocket': enableWS
            }
        
        r = requests.post(self.url + '/config', data = json.dumps(d))
        response = json.loads(r.text)

        return response

    def msgChain(self, s):
        c = [
            {
                'type': 'Plain',
                'text': s
            },
        ]
        return c

class Receive:
    def __init__(self, sessionKey, msgRouter):
        self.sessionKey = sessionKey
        self.port = port
        self.msgRouter = msgRouter

    def on_message(self, message):
        response = json.loads(message)
        self.msgRouter.onReceived(response)

    def on_error(self, error):
        print('web_api/api_mirai_http.py::Receive::on_error:', error)

    def on_close(self):
        print('websocket closed')

    def startListen(self):
        #websocket.enableTrace(True)
        ws = websocket.WebSocketApp(ws_url + '/all?sessionKey=' + self.sessionKey,
        on_message=self.on_message,
        on_error=self.on_error,
        on_close=self.on_close)
        
        print('start receiving')
        ws.run_forever()

File: liqibot2/src/web_api/api_mirai_http.cpp
﻿#pragma once
#include <json/json.h>

#include "api_mirai_http.h"
#include "../Requests.h"
#include "../WebSocket.h"
#include "../Log.h"

QQApi::QQApi()
{
	//默认config
	std::string conf = R""(
		{
			"qq": 2155679839,
			"host": "localhost",
			"port": 8080,
			"verifyKey": "123456789"
		}
		)"";

	config = parseJson(conf);
	loadConfig();
	saveConfig();

	qq = config["qq"].asInt64();
	host = config["host"].asString();
	port = config["port"].asInt();
	verifyKey = config["verifyKey"].asString();
	http_url = std::string("http://") + host + ":" + std::to_string(port);
	ws_url = std::string("ws://") + host + ":" + std::to_string(port);
}

QQApi::~QQApi()
{
}

void QQApi::saveConfig()
{
	std::ofstream config_file(config_path);
	config_file << dumpsJson(config);
	config_file.close();
}

void QQApi::loadConfig()
{
	try
	{
		config = loadJson(config_path);
	}
	catch (const std::string& e)//解析失败，使用默认config
	{
		std::cout << "Load " + config_path + " failed.\n"
			<< e
			<< "Using default config" << "\n\n";
	}
}

Json::Value QQApi::about()
{
	Requests r = Requests::get(http_url + "/about");

	return parseJson(r.text);
}

int QQApi::connect()
{
	if (verify() == 0)
	{
		if (bind() == 0)
		{
			return 0;
		}
	}
	return -1;
}

int QQApi::verify()
{
	Json::Value d;
	d["verifyKey"] = verifyKey;

	Requests r = Requests::post(http_url + "/verify", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		if (res["code"].asInt() == 0)
		{
			sessionKey = res["session"].asString();
		}

		return res["code"].asInt();
	}

	return -1;
}

int QQApi::bind()
{
	Json::Value d;
	d["sessionKey"] = sessionKey;
	d["qq"] = this->qq;

	Requests r = Requests::post(http_url + "/bind", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		if (res["code"].asInt() == 0)
		{
			std::cout << "bind success" << "\n";
			return 0;
		}
		return res["code"].asInt();
	}

	return -1;
}

int QQApi::release()
{
	Json::Value d;
	d["sessionKey"] = sessionKey;
	d["qq"] = qq;

	Requests r = Requests::post(http_url + "/release", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		if (res["code"].asInt() == 0)
		{
			std::cout << "release success" << "\n";
			return 0;
		}
		return res["code"].asInt();
	}

	return -1;
}

int64 QQApi::sendMessage(Member member, int64 quote, std::string s)
{
	return sendMessage(member, quote, MessageChain::fromString(s));
}

int64 QQApi::sendMessage(Member member, int64 quote, MessageChain msgChain)
{
	if (member.group.id == 0)
	{
		return sendFriendMessage(member, quote, msgChain);
	}
	else
	{
		return sendGroupMessage(member.group, quote, msgChain);
	}
}

int64 QQApi::sendFriendMessage(int64 qq, int64 quote, MessageChain msgChain)
{
	Json::Value d;
	d["sessionKey"] = sessionKey;
	d["qq"] = qq;
	d["messageChain"] = msgChain.toJson();
	if (quote != 0)
	{
		d["quote"] = quote;
	}

	Requests r = Requests::post(http_url + "/sendFriendMessage", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		if (res["code"].asInt() == 0)
		{
			Log::add("send", dumpsJson(d, false), this->qq);
			return res["messageId"].asInt64();
		}
		return res["code"].asInt();
	}

	return -1;
}

int64 QQApi::sendFriendMessage(Member member, int64 quote, MessageChain msgChain)
{
	return sendFriendMessage(member.id, quote, msgChain);
}

int64 QQApi::sendTempMessage(int64 group, int64 qq, int64 quote, MessageChain msgChain)
{
	Json::Value d;
	d["sessionKey"] = sessionKey;
	d["qq"] = qq;
	d["group"] = group;
	d["messageChain"] = msgChain.toJson();
	if (quote != 0)
	{
		d["quote"] = quote;
	}

	Requests r = Requests::post(http_url + "/sendTempMessage", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		if (res["code"].asInt() == 0)
		{
			Log::add("send", dumpsJson(d, false), this->qq);
			return res["messageId"].asInt64();
		}
		return res["code"].asInt();
	}

	return -1;
}

int64 QQApi::sendTempMessage(Group group, Member member, int64 quote, MessageChain msgChain)
{
	return sendTempMessage(group.id, member.id, quote, msgChain);
}

int64 QQApi::sendGroupMessage(int64 group, int64 quote, MessageChain msgChain)
{
	Json::Value d;
	d["sessionKey"] = sessionKey;
	d["group"] = group;
	d["messageChain"] = msgChain.toJson();
	if (quote != 0)
	{
		d["quote"] = quote;
	}

	//std::cout << dumpsJson(d, false) << std::endl;
	Requests r = Requests::post(http_url + "/sendGroupMessage", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		if (res["code"].asInt() == 0)
		{
			Log::add("send", dumpsJson(d, false), this->qq);
			return res["messageId"].asInt64();
		}
		else
		{
			std::cout << res["msg"] << std::endl;
		}
		return res["code"].asInt();
	}

	return -1;
}

int64 QQApi::sendGroupMessage(Group group, int64 quote, MessageChain msgChain)
{
	return sendGroupMessage(group.id, quote, msgChain);
}

int64 QQApi::sendImageMessage()
{
	return 0;
}

int64 QQApi::uploadImage()
{
	return 0;
}

int QQApi::recall(int64 msgId)
{
	Json::Value d;
	d["sessionKey"] = sessionKey;
	d["target"] = msgId;
	
	Requests r = Requests::post(http_url + "/recall", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		if (res["code"].asInt() == 0)
		{
			Log::add("send", dumpsJson(d, false), this->qq);
			return 0;
		}
		return -1;
	}

	return -1;
}

std::vector<Member> QQApi::getFriendList()
{
	Json::Value d;
	d["sessionKey"] = sessionKey;

	Requests r = Requests::get(http_url + "/friendList", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		std::vector<Member> members;
		for (int i = 0; i < res.size(); i++)
		{
			Member m;
			m.id = res[i]["id"].asInt64();
			m.nickname = res[i]["nickname"].asString();
			m.remark = res[i]["remark"].asString();
			members.push_back(m);
		}
		return members;
	}

	return std::vector<Member>();
}

std::vector<Group> QQApi::getGroupList()
{
	Json::Value d;
	d["sessionKey"] = sessionKey;

	Requests r = Requests::get(http_url + "/groupList", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		std::vector<Group> groups;
		for (int i = 0; i < res.size(); i++)
		{
			Group g;
			g.id = res[i]["id"].asInt64();
			g.name = res[i]["name"].asString();
			switch (stringhash_run_time(res[i]["permission"].asString().c_str()))
			{
			case "MEMBER"_hash:
				g.permission = Group::MEMBER;
				break;
			case "ADMINISTRATOR"_hash:
				g.permission = Group::ADMINISTRATOR;
				break;
			case "OWNER"_hash:
				g.permission = Group::OWNER;
				break;
			default:
				break;
			}
			
			groups.push_back(g);
		}
		return groups;
	}

	return std::vector<Group>();
}

std::vector<Member> QQApi::getMemberList()
{
	Json::Value d;
	d["sessionKey"] = sessionKey;

	Requests r = Requests::get(http_url + "/friendList", dumpsJson(d));
	if (r.code == 200)
	{
		Json::Value res = parseJson(r.text);
		std::vector<Member> members;
		for (int i = 0; i < res.size(); i++)
		{
			Member m;
			m.id = res[i]["id"].asInt64();
			m.memberName = res[i]["memberName"].asString();
			switch (stringhash_run_time(res[i]["permission"].asString().c_str()))
			{
			case "MEMBER"_hash:
				m.permission = Member::MEMBER;
				break;
			case "ADMINISTRATOR"_hash:
				m.permission = Member::ADMINISTRATOR;
				break;
			case "OWNER"_hash:
				m.permission = Member::OWNER;
				break;
			default:
				break;
			}

			Group g;
			g.id = res[i]["id"]["group"]["id"].asInt64();
			g.name = res[i]["id"]["group"]["name"].asString();
			switch (stringhash_run_time(res[i]["id"]["group"]["permission"].asString().c_str()))
			{
			case "MEMBER"_hash:
				g.permission = Group::MEMBER;
				break;
			case "ADMINISTRATOR"_hash:
				g.permission = Group::ADMINISTRATOR;
				break;
			case "OWNER"_hash:
				g.permission = Group::OWNER;
				break;
			default:
				break;
			}

			m.group = g;
			members.push_back(m);
		}
		return members;
	}

	return std::vector<Member>();
}

int QQApi::muteAll(Group group)
{
	return 0;
}

int QQApi::unmuteAll(Group group)
{
	return 0;
}

int QQApi::mute(Group group, Member member, int sec)
{
	return 0;
}

int QQApi::unmute(Group group, Member member)
{
	return 0;
}

int QQApi::kick()
{
	return 0;
}

int QQApi::quit(Group group)
{
	return 0;
}

int QQApi::setGroupConfig()
{
	return 0;
}

Json::Value QQApi::getGroupConfig(int64 group)
{
	return Json::Value();
}

Json::Value QQApi::getGroupConfig(Group group)
{
	return Json::Value();
}

int QQApi::setMemberInfo()
{
	return 0;
}

Json::Value QQApi::getMemberInfo(int64 group, int64 qq)
{
	return Json::Value();
}

Json::Value QQApi::getMemberInfo(Group group, Member member)
{
	return Json::Value();
}

int QQApi::setSessionConfig(int cacheSize, bool enableWS)
{
	return 0;
}

Json::Value QQApi::getSessionConfig()
{
	return Json::Value();
}

void QQApi::startListen(std::function<void(std::string, QQApi* qqApi)> onReceived)
{
	this->onReceived = onReceived;
	auto ws = std::make_shared<WebSocket>(ws_url + "/all?" + "verifyKey=" + verifyKey + "&qq=" + std::to_string( qq),
		[this]() {this->onHandshake(); },
		[this](std::string s) {this->onMessage(s); }, 
		[this](std::string s) {this->onErr(s); }, 
		[this]() {this->onClose(); }
	);
	ws->start();
	ws->run();//阻塞的
}

void QQApi::onHandshake()
{
}

void QQApi::onMessage(std::string s)
{
	std::cout << s << "\n";
	
	(onReceived)(s, this);
}

void QQApi::onErr(std::string s)
{
}

void QQApi::onClose()
{
}

